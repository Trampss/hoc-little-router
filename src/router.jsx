import React from 'react'

const getDisplayName = Component => `router(${
  Component.displayName
  || Component.name
  || (Component.constructor && Component.constructor.name)
  || 'Unknown'
})`

const isRouteFound = title => result => result && [].concat(title).includes(result.title)

const hoc = (title, options) => Component => class extends React.Component {
  static displayName = getDisplayName(Component)

  static contextTypes = {
    store: () => null, // this is to avoid importing prop-types
  }

  constructor(props, context) {
    super(props, context)

    this.state = { show: false }
  }

  componentWillMount() {
    const { store } = this.context

    // subscribe
    this.unsubscribe = store.subscribe(() => {
      this.toShow()
    })

    // run in once
    this.toShow()
  }

  componentWillUnmount() {
    this.unsubscribe()
  }

  toShow = () => {
    const { store } = this.context

    const state = store.getState()

    if (!state.router || !state.router.result) {
      // eslint-disable-next-line no-console
      console.error('/HOC/ Router | There is no route found in `state.router.result`')
      return
    }

    // Absolute mode, we are looking in top level only
    let { result } = state.router
    if (options && options.absolute) {
      const show = isRouteFound(title)(result)
      if (show !== this.state.show) { // eslint-disable-line react/destructuring-assignment
        this.setState(innerState => ({ ...innerState, show }))
      }

      return
    }

    // Either way we are looking top down the result tree
    let show = isRouteFound(title)(result)
    while (result && !show) {
      result = result.parent
      show = isRouteFound(title)(result)
    }

    if (show !== this.state.show) { // eslint-disable-line react/destructuring-assignment
      this.setState(innerState => ({ ...innerState, show }))
    }
  }

  render() {
    const { show } = this.state

    if (!show) return null

    return <Component {...this.props} />
  }
}

hoc.absolute = (title, options) => hoc(title, { ...options, absolute: true })

export default hoc
