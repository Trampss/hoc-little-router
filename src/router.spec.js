/* eslint-disable
 import/no-extraneous-dependencies,
 react/jsx-filename-extension,
 react/prop-types
 */
/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import sinon from 'sinon'
import router from './router'

const Component = props => <div>{JSON.stringify(props)}</div>

const store = createStore(() => ({
  router: {
    result: {
      title: 'TITLE_1',
      parent: {
        title: 'TITLE_2',
      },
    },
  },
}))

const snap = (title, options, st = store) => {
  const Decorated = router(title, options)(Component)

  const component = renderer.create(
    <Provider store={st}>
      <Decorated prop1="prop1" show />
    </Provider>,
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
}

describe('Router -redux-little-router- HOC', () => {
  it('should print Component on first level', () => snap('TITLE_1'))
  it('should not print Component', () => snap('TITLE_42'))
  it('should print Component on second level', () => snap('TITLE_2'))
  it('should not print Component on second level (absolute mode)', () => snap('TITLE_2', { absolute: true }))

  describe('Errors', () => {
    let spy
    beforeEach(() => { spy = sinon.stub(console, 'error') })
    afterEach(() => { spy.restore() })

    const snapError = (state) => {
      snap('TITLE_1', undefined, createStore(() => state))
      expect(spy.calledOnce).toBe(true)
      expect(spy.getCall(0).args[0]).toEqual('/HOC/ Router | There is no route found in `state.router.result`')
    }

    it('should print an error when router reducer is not found', () => snapError({}))
    it('should print an error when no result is found', () => snapError({ router: {} }))
  })
})

/* eslint-enable
 import/no-extraneous-dependencies,
 react/jsx-filename-extension,
 react/prop-types
 */
