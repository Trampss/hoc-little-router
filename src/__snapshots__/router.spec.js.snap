exports[`Router -redux-little-router- HOC Errors should print an error when no result is found 1`] = `null`;

exports[`Router -redux-little-router- HOC Errors should print an error when router reducer is not found 1`] = `null`;

exports[`Router -redux-little-router- HOC should not print Component 1`] = `null`;

exports[`Router -redux-little-router- HOC should not print Component on second level (absolute mode) 1`] = `null`;

exports[`Router -redux-little-router- HOC should print Component on first level 1`] = `
<div>
  {\"prop1\":\"prop1\",\"show\":true}
</div>
`;

exports[`Router -redux-little-router- HOC should print Component on second level 1`] = `
<div>
  {\"prop1\":\"prop1\",\"show\":true}
</div>
`;
